### PR TITLE
feat(jellyfish-api-core, jellyfish-transaction): add evmTx 

### PR DESCRIPTION
**Pull request by lykalabrada**
#### What this PR does / why we need it:
Adds support for creation of EVM transaction.

#### Which issue(s) does this PR fixes?:

Fixes #

### CODE REVIEW

File: packages/testcontainers/src/containers/RegTestContainer/index.ts [hunk 1]
```diff
@@ -55,7 +56,7 @@ export class RegTestContainer extends DeFiDContainer {
     return cmds
   }
 
-  async getNewAddress (label: string = '', addressType: 'legacy' | 'p2sh-segwit' | 'bech32' | string = 'bech32'): Promise<string> {
+  async getNewAddress (label: string = '', addressType: 'legacy' | 'p2sh-segwit' | 'bech32' | 'eth' | string = 'bech32'): Promise<string> {
```
Comment (infinia-yzl) [inline packages/testcontainers/src/containers/RegTestContainer/index.ts:59]: Actually, you may start using `NativeChainContainer`.

https://github.com/BirthdayResearch/jellyfishsdk/blob/afed8a45e3a7b768478ff368b695ceea0c44201a/packages/testcontainers/src/containers/NativeChainRpc.ts#L120C7-L122

Refer to the tests for how to use / implement:
https://github.com/BirthdayResearch/jellyfishsdk/blob/main/packages/testcontainers/__tests__/containers/NativeChainContainer/Masternode.test.ts

File: packages/jellyfish-api-core/__tests__/category/evm/evmTx.test.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+      // data: '0x00'
+    })
+    await container.generate(1)
+
+    const blockHash: string = await client.blockchain.getBestBlockHash()
+    const txs = await client.blockchain.getBlock(blockHash, 1)
+    expect(txs.tx[1]).toStrictEqual(evmTxHash)
+  })
+
+  it('should fail creation of evmtx when balance is not enough', async () => {
```
Comment (pierregee) [inline packages/jellyfish-api-core/__tests__/category/evm/evmTx.test.ts]: Can add more fine-grained tests for non-happy paths -> invalid nonce, gas price, gasLimit, value

File: packages/jellyfish-transaction/src/script/crypto.ts [hunk 3]
```diff
@@ -18,6 +18,15 @@ export class OP_SHA1 extends StaticCode {
   }
 }
 
+/**
+ * The input is hashed using SHA-3.
+ */
+export class OP_SHA3 extends StaticCode {
```
Comment (pierregee) [inline packages/jellyfish-transaction/src/script/crypto.ts:24]: Is this being used by the `evmtx` internally? ain script [here](https://github.com/DeFiCh/ain/blob/cb38e0fc35920e34b2e8451128b3724d590ffd59/src/script/script.h#L167), just want to make sure if it is needed

**Review by pierregee — CHANGES_REQUESTED** [review]

File: packages/jellyfish-api-core/__tests__/category/evm/evmTx.test.ts [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+      from: ethAddress,
+      to: toEthAddress,
+      nonce: 0,
+      gasPrice: 21,
+      gasLimit: 21000,
+      value: new BigNumber(10)
+    })).rejects.toThrow(RpcApiError)
+  })
+
+  it('should fail creation of evmtx when ethereum address is not valid', async () => {
```
Comment (pierregee) [inline packages/jellyfish-api-core/__tests__/category/evm/evmTx.test.ts]: can add test 1 test for invalid `from` and another for invalid `to`

File: packages/jellyfish-api-core/__tests__/category/evm/evmTx.test.ts [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    // Create EVM transaction and submit to local node
+    const toEthAddress = await container.call('getnewaddress', ['', 'eth'])
+    const evmTxHash = await client.evm.evmtx({
+      from: ethAddress,
+      to: toEthAddress,
+      nonce: 0,
+      gasPrice: 21,
+      gasLimit: 21000,
+      value: new BigNumber(1)
+      // data: '0x00'
```
Comment (fullstackninja864) [inline packages/jellyfish-api-core/__tests__/category/evm/evmTx.test.ts]: do we need this ?
 lets add one test case with data field if possible

File: packages/jellyfish-transaction/src/script/crypto.ts (hunk 3 above)
Comment (lykalabrada) [reply packages/jellyfish-transaction/src/script/crypto.ts:24]: Nope, not used by evmtx. just thought it would be nice to keep jellyfish updated

Comment (lykalabrada) [reply packages/jellyfish-transaction/src/script/crypto.ts:24]: @canonbrother can we keep this `OP_SHA3` here? Or should I remove? 🤔

Comment (canonbrother) [reply packages/jellyfish-transaction/src/script/crypto.ts:24]: hmm... i think its fine to keep there

File: packages/jellyfish-api-core/src/category/evm.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+  /**
+   * Creates an EVM transaction submitted to local node and network
+   * @param {string} from
+   * @param {number} nonce
+   * @param {number} gasPrice
+   * @param {number} gasLimit
+   * @param {string} to
+   * @param {BigNumber} value
+   * @param {string} data
```
Comment (canonbrother) [inline packages/jellyfish-api-core/src/category/evm.ts]: optional param in jsdoc
```suggestion
   * @param {string} [data]
```

File: packages/jellyfish-transaction/__tests__/script/dftx/dftx_evm/EvmTx.test.ts [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    header: '6a4c744466547839', // OP_RETURN(6a) OP_PUSHDATA1(4c) (length 74) CDfTx.SIGNATURE(44665478) CEvmTx.OP_CODE(39)
+    data: '6ef86c018504e3b2920082520894585bd64cf6574abf77f216efc894940e87cad5b1880de0b6b3a76400008026a035e16fa88aa4a1d01990c3b6acb0e6d869fe4d5960992490814f65e063ba3ed7a0401a72637c21501313e1a5c8b03cd1031ea7823085c6a8f6d388caf078027d7b',
+    evmTx: {
+      raw: 'f86c018504e3b2920082520894585bd64cf6574abf77f216efc894940e87cad5b1880de0b6b3a76400008026a035e16fa88aa4a1d01990c3b6acb0e6d869fe4d5960992490814f65e063ba3ed7a0401a72637c21501313e1a5c8b03cd1031ea7823085c6a8f6d388caf078027d7b'
+    }
+  }
+]
+
+describe.each(evmTxData)('should craft and compose dftx',
+  ({ header, evmTx, data }: { header: string, data: string, evmTx: EvmTx }) => {
```
Comment (canonbrother) [inline packages/jellyfish-transaction/__tests__/script/dftx/dftx_evm/EvmTx.test.ts]: :thinking: 
```suggestion
  ({ header, data, evmTx }: { header: string, data: string, evmTx: EvmTx }) => {
```

**Review by canonbrother — COMMENTED** [review]
overall good tho

can try passing a "non-hex" data in evmtx??

**Comment by lykalabrada** [comment]
> overall good tho
> 
> can try passing a "non-hex" data in evmtx??

Added 🙏 
https://github.com/BirthdayResearch/jellyfishsdk/blob/268df1e28f880bba6d6827a60cdc415912f4d636/packages/jellyfish-api-core/__tests__/category/evm/evmTx.test.ts#L98

**Review by canonbrother — APPROVED** [review]

**Review by thedoublejay — APPROVED** [review]

**Review by pierregee — APPROVED** [review]

File: packages/jellyfish-transaction/src/script/crypto.ts [hunk 8]
```diff
@@ -18,6 +18,15 @@ export class OP_SHA1 extends StaticCode {
...
+  constructor () {
+    super(0xc0, 'OP_SHA3')
```
Comment (fuxingloh) [inline packages/jellyfish-transaction/src/script/crypto.ts:26]: Just checking. Does this cause any collision with existing OP_CODE from Bitcoin Core.

**Review by fuxingloh — APPROVED** [review]

File: packages/jellyfish-transaction/src/script/crypto.ts (hunk 8 above)
Comment (lykalabrada) [reply packages/jellyfish-transaction/src/script/crypto.ts:26]: Don’t think so, it’s not used at all by bitcoin

**Pull request merged**